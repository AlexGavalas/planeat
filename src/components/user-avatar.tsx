import { Avatar } from '@mantine/core';
import Image from 'next/image';

export const UserAvatar = ({ src }: { src: string }) => {
    return (
        <Avatar>
            <Image
                src={src}
                alt="user avatar image"
                layout="intrinsic"
                width={50}
                height={50}
            />
        </Avatar>
    );
};
