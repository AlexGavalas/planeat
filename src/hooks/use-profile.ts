import { type Session } from 'next-auth';
import { signOut } from 'next-auth/react';
import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import { useMutation, useQuery, useQueryClient } from 'react-query';

import { type User } from '~types/user';
import { showSuccessNotification } from '~util/notification';

import { useUser } from './use-user';

type MutationProps = {
    isDiscoverable?: boolean;
    height?: number;
    targetWeight?: number;
    language?: string;
    hasCompletedOnboarding?: boolean;
    silent?: boolean;
};

type UseProfile = () => {
    profile?: User;
    isFetching: boolean;
    updateProfile: (params: MutationProps) => void;
    user: Session['user'];
    deleteProfile: () => void;
    isDeleting: boolean;
};

export const useProfile: UseProfile = () => {
    const router = useRouter();
    const queryClient = useQueryClient();
    const { t } = useTranslation();
    const user = useUser();

    const { data: profile, isFetching } = useQuery(
        ['user'],
        async () => {
            const response = await fetch('/api/v1/user');

            const { data } = (await response.json()) as { data?: User };

            return data;
        },
        {
            enabled: Boolean(user?.email),
        },
    );

    const { mutate: updateProfile, isLoading } = useMutation(
        async ({
            isDiscoverable,
            height,
            targetWeight,
            language,
            hasCompletedOnboarding,
        }: MutationProps) => {
            const response = await fetch('/api/v1/user', {
                method: 'PATCH',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    isDiscoverable,
                    height,
                    targetWeight,
                    language,
                    hasCompletedOnboarding,
                }),
            });

            if (!response.ok) {
                // TODO: Handle error
            }
        },
        {
            onSuccess: async (_, { language, silent }) => {
                if (!silent) {
                    showSuccessNotification({
                        title: t('notification.success.title', {
                            lng: language,
                        }),
                        message: t('notification.success.message', {
                            lng: language,
                        }),
                    });
                }

                await queryClient.invalidateQueries(['user']);
            },
        },
    );

    const { mutate: deleteProfile } = useMutation(
        async () => {
            const response = await fetch('/api/v1/user', {
                method: 'DELETE',
            });

            if (!response.ok) {
                // TODO: Handle error
            }
        },
        {
            onSuccess: async () => {
                showSuccessNotification({
                    title: t('notification.success.title'),
                    message: t('notification.success.message'),
                });

                queryClient.clear();

                await signOut();

                await router.push('/');
            },
        },
    );

    return {
        profile,
        isFetching,
        updateProfile,
        user,
        deleteProfile,
        isDeleting: isLoading,
    };
};
