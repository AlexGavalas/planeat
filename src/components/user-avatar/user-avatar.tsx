import { Avatar } from '@mantine/core';
import Image from 'next/image';

import { useProfile } from '~hooks/use-profile';

export const UserAvatar = () => {
    const { user } = useProfile();

    if (!user?.image || !user.name) {
        return null;
    }

    return (
        <Avatar radius="xl" size="sm">
            <Image src={user.image} alt={user.name} width={32} height={32} />
        </Avatar>
    );
};
