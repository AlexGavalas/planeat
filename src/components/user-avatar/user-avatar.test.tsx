import { useProfile } from '~hooks/use-profile';
import { renderWithUser, screen } from '~test/utils';

import { UserAvatar } from './user-avatar';

jest.mock('~hooks/use-profile');

describe('<UserAvatar />', () => {
    const defaultProfile = {
        deleteProfile: jest.fn(),
        isDeleting: false,
        isFetching: false,
        profile: undefined,
        updateProfile: jest.fn(),
        user: undefined,
    } satisfies ReturnType<typeof useProfile>;

    describe('when the user is not logged in', () => {
        beforeAll(() => {
            jest.mocked(useProfile).mockReturnValue({
                ...defaultProfile,
                user: undefined,
            });
        });

        it('does not render', () => {
            renderWithUser(<UserAvatar />);

            expect(screen.queryByRole('img')).toBeNull();
        });
    });

    describe('when the user has no image', () => {
        beforeAll(() => {
            jest.mocked(useProfile).mockReturnValue({
                ...defaultProfile,
                user: { image: undefined, name: 'Test User' },
            });
        });

        it('does not render', () => {
            renderWithUser(<UserAvatar />);

            expect(screen.queryByRole('img')).toBeNull();
        });
    });

    describe('when the user has no name', () => {
        beforeAll(() => {
            jest.mocked(useProfile).mockReturnValue({
                ...defaultProfile,
                user: { image: 'http://planeat/image', name: undefined },
            });
        });

        it('does not render', () => {
            renderWithUser(<UserAvatar />);

            expect(screen.queryByRole('img')).toBeNull();
        });
    });

    describe('when the user is logged in', () => {
        beforeAll(() => {
            jest.mocked(useProfile).mockReturnValue({
                ...defaultProfile,
                user: { image: 'http://planeat/image', name: 'Test User' },
            });
        });

        it('render the avatar', () => {
            const { container } = renderWithUser(<UserAvatar />);

            expect(screen.queryByRole('img')).toBeInTheDocument();
            expect(container).toMatchSnapshot();
        });
    });
});
