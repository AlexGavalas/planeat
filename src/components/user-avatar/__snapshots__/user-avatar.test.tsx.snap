// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAvatar /> when the user is logged in render the avatar 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m-f85678b6 mantine-Avatar-root"
    data-size="sm"
    style="--avatar-size: var(--avatar-size-sm); --avatar-radius: var(--mantine-radius-xl);"
  >
    <span
      class="m-104cd71f mantine-Avatar-placeholder"
    >
      <img
        alt="Test User"
        data-nimg="1"
        decoding="async"
        height="32"
        loading="lazy"
        src="/_next/image?url=http%3A%2F%2Fplaneat%2Fimage&w=64&q=75"
        srcset="/_next/image?url=http%3A%2F%2Fplaneat%2Fimage&w=32&q=75 1x, /_next/image?url=http%3A%2F%2Fplaneat%2Fimage&w=64&q=75 2x"
        style="color: transparent;"
        width="32"
      />
    </span>
  </div>
</div>
`;
